### PR TITLE
feat: adapting spiaccesscheck to remotesecrets

### PR DESCRIPTION
**Pull request by xbaran4**
### What does this PR do?
Extends SPIAccessCheck's functionality.

The CheckAccess first tries to find SPIAccessToken to construct a client and determine if a repository is accessible.
If no SPIAccessToken is found, this PR comes into play. We find all RemoteSecrets in the same namespace as SPIAccessCeck with the label `appstudio.redhat.com/sp.host` and filter this list of RemoteSecrets so that only those that satisfy these conditions are kept:
- has Data Obtained Condition
- has `secret.type == kubernetes.io/basic-auth`
- has target in the same namespace as SPIAccessCheck (and the RemoteSecret itself)

From these, we preferentially pick an RS that contains the repository name (from SPIAccessCheck ) in  `appstudio.redhat.com/sp.repo`.

We then use the RemoteSecret's Secret to construct an authenticated client and determine whether repository is accessible.

The SPIAccessCheck CRD is extended with the field:
```yaml
status:
  credentials: 
    remoteSecret: __name of remote secret__
    secret: __name of secret__
```

#### GitHub
image

#### Quay
image

#### GitLab
image

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-564

### How to test this PR?

Deploy SPI and RemoteSecret
SPIO_IMG=quay.io/pbaran/spio:SVPI-564-spiaccesscheck

### RUN SCRIPT
https://github.com/xbaran4/spi-scripts/blob/master/test-access.sh
I have created a script where you can fill out your tokens and names of public/private repositories, and it tests three scenarios for each service provider:
1. SPIAccessCheck with no tokens for a public repository
2. SPIAccessCheck with SPIAccessToken for a private repository
3. SPIAccessCheck with RemoteSecret for a private repository 

#### GitHub
1. Create Github Personal Access Token (classic) (https://github.com/settings/tokens) `repo` scope should suffice
4. Create (GitHub) RemoteSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret
  namespace: default
  labels:
    appstudio.redhat.com/sp.host: github.com
spec:
  secret:
    type: kubernetes.io/basic-auth
    generateName: secret-from-remote-
  targets:
    - namespace: default
EOF
```
5. Create UploadSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret
type: kubernetes.io/basic-auth
stringData:
  password: "_your GitHub PAT_"
EOF
```
6. Create SPIAccessCheck with repoUrl referencing your private repo:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: SPIAccessCheck
metadata:
  name: spiaccesscheck-sample
  namespace: default
spec:
  repoUrl: "https://github.com/xbaran4/private"
EOF
```
7. You should see SPIAccessCheck similar to the state in the screenshot

#### Quay
1. Create Quay robot token. Make sure it has permission (`write`) to access your private registry
2. Create (Quay) RemoteSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret
  namespace: default
  labels:
    appstudio.redhat.com/sp.host: quay.io
spec:
  secret:
    type: kubernetes.io/basic-auth
    generateName: secret-from-remote-
  targets:
    - namespace: default
EOF
```
3. Create UploadSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret
type: kubernetes.io/basic-auth
stringData:
  username: "pbaran+test" #replace with your robot name
  password: "_your robot token_"
EOF
```
4. Create SPIAccessCheck with repoUrl referencing your private repo:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: SPIAccessCheck
metadata:
  name: spiaccesscheck-sample
  namespace: default
spec:
  repoUrl: "https://quay.io/pbaran/private-test" # replace with your private reg
EOF
```
5. You should see SPIAccessCheck similar to the state in the screenshot

#### GitLab
1. Create GitLab Personal Access Token (https://gitlab.com/-/profile/personal_access_tokens) `api` scope should suffice
2. Create (GitLab) RemoteSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret
  namespace: default
  labels:
    appstudio.redhat.com/sp.host: gitlab.com
spec:
  secret:
    type: kubernetes.io/basic-auth
    generateName: secret-from-remote-
  targets:
    - namespace: default
EOF
```
3. Create UploadSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret
type: kubernetes.io/basic-auth
stringData:
  password: "_your GitLab PAT_"
EOF
```
4. Create SPIAccessCheck with repoUrl referencing your private repo:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: SPIAccessCheck
metadata:
  name: spiaccesscheck-sample
  namespace: default
spec:
  repoUrl: "https://gitlab.com/xbaran4/test"
EOF
```
5. You should see SPIAccessCheck similar to the state in the screenshot

```
quay.io/redhat-appstudio/pull-request-builds:spi-controller-#
quay.io/redhat-appstudio/pull-request-builds:spi-oauth-#
```

### CODE REVIEW

File: pkg/serviceprovider/tokenfilter.go [hunk 1]
```diff
@@ -52,3 +54,27 @@ func NewFilter(policy config.TokenPolicy, exactTokenFilter TokenFilter) TokenFil
...
+
+func (f RemoteSecretFilterFunc) Matches(ctx context.Context, matchable Matchable, remoteSecret *v1beta1.RemoteSecret) bool {
+	return f(ctx, matchable, remoteSecret)
+}
+
+var _ RemoteSecretFilter = (RemoteSecretFilterFunc)(nil)
+
+// DefaultRemoteSecretFilterFunc filters RemoteSecret based on three conditions: the SecretData must be obtained,
+// type of the secret must SecretTypeBasicAuth, and the RemoteSecret must have target in the same (local) namespace as matchable.
+var DefaultRemoteSecretFilterFunc = RemoteSecretFilterFunc(func(ctx context.Context, matchable Matchable, remoteSecret *v1beta1.RemoteSecret) bool {
```
[the remark this answers is not in the document]
Comment (xbaran4) [reply pkg/serviceprovider/tokenfilter.go:75]: sounds reasonable. will do.

File: pkg/serviceprovider/gitlab/gitlab.go [hunk 2]
```diff
@@ -41,12 +41,19 @@ var unsupportedScopeError = errors.New("unsupported scope for GitLab")
...
 var unsupportedUserWritePermissionError = errors.New("user write permission is not supported by GitLab")
 var probeNotImplementedError = errors.New("gitLab probe not implemented")
+var unexpectedStatusCodeError = errors.New("unexpected status code from GitLab API")
 
 var publicRepoMetricConfig = serviceprovider.CommonRequestMetricsConfig(config.ServiceProviderTypeGitLab, "fetch_public_repo")
 var fetchRepositoryMetricConfig = serviceprovider.CommonRequestMetricsConfig(config.ServiceProviderTypeGitLab, "fetch_single_repo")
 
 var _ serviceprovider.ServiceProvider = (*Gitlab)(nil)
 
+var defaultAccessCheckStatus = api.SPIAccessCheckStatus{
```
[the remark this answers is not in the document]
Comment (xbaran4) [reply pkg/serviceprovider/gitlab/gitlab.go:51]: ahh, thats a leftover from older commits. I think here it doesn't make sense. will fix

File: pkg/serviceprovider/lookup.go [hunk 3]
```diff
@@ -39,32 +51,34 @@ type GenericLookup struct {
...
 	// MetadataCache is an abstraction used for storing/fetching the metadata of tokens
 	MetadataCache *MetadataCache
-	// RepoHostParser is a function that extracts the host from the repoUrl
-	RepoHostParser RepoHostParser
+	// RepoUrlParser is a function that parses URL from the repoUrl
+	RepoUrlParser RepoUrlParser
 }
 
-type RepoHostParser func(url string) (string, error)
+type RepoUrlParser func(url string) (*url.URL, error)
```
[the remark this answers is not in the document]
Comment (xbaran4) [reply pkg/serviceprovider/lookup.go:65]: I'm actually more inclined to make `RepoUrlFromSchemalessString` a util function instead, and remove the Parser from the `GenericLookup`'s field.

I feel like `RepoUrlFromSchemalessString` can be used everywhere in our current situation.

I did not do it as I wanted to limit the PR size as much as possible but let me know if you would agree with such a change.

File: pkg/serviceprovider/clientbuilder.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import "context"
+
+type AuthenticatedClientBuilder[C any] interface {
+	CreateAuthenticatedClient(ctx context.Context, credentials Credentials) (*C, error)
+}
+
+type Credentials struct {
+	Username         string
+	Token            string
+	SourceObjectName string
```
[the remark this answers is not in the document]
Comment (xbaran4) [reply pkg/serviceprovider/clientbuilder.go:26]: If we do not know/have the need for the `CredentialsReference` in `SPIAccessCheckStatus` I think my preference would be to just remove the `CredentialsReference` and `SourceObjectName` from the `Credentials` struct.|

In the future we can decide how we want to extend the status and as a consequence what should be in the `Credentials` struct. `OwnerReference` might be one of the possibilities.

File: api/v1beta1/spiaccesscheck_types.go [hunk 5]
```diff
@@ -23,6 +23,16 @@ import (
...
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
+const (
+	// RSServiceProviderHostLabel is supposed to be present on RemoteSecret that contains credentials which can be used
+	// for SPIAccessCheck. The value represent service provider hostname.
+	RSServiceProviderHostLabel = "appstudio.redhat.com/sp.host"
+
+	// RSServiceProviderRepositoryAnnotation provides additional information on top of RSServiceProviderHostLabel.
+	// The value represents comma-separated names of repositories for which the credentials can be used.
+	RSServiceProviderRepositoryAnnotation = "appstudio.redhat.com/sp.repo"
```
[the remark this answers is not in the document]
Comment (xbaran4) [reply api/v1beta1/spiaccesscheck_types.go:33]: I have no preference for this one. If you and others prefer the change, let's do it.

File: pkg/serviceprovider/github/githubclientbuilder.go [hunk 6]
```diff
@@ -33,6 +35,14 @@ type githubClientBuilder struct {
 	tokenStorage tokenstorage.TokenStorage
 }
 
+func (g *githubClientBuilder) CreateAuthenticatedClient(ctx context.Context, credentials serviceprovider.Credentials) (*github.Client, error) {
```
Comment (mshaposhnik) [inline pkg/serviceprovider/github/githubclientbuilder.go:38]: What's the idea of using pointer type recievers there?

Comment (xbaran4) [reply pkg/serviceprovider/github/githubclientbuilder.go:38]: yes, good point. I will refactor it for the main types like `Github` and `Gitlab`, but not for structs like `downloadFileCapability` as that calls a non-interfaced function (`createGitlabAuthClient`) and I do not want to refactor it in this PR.

File: pkg/serviceprovider/rsfilter.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+func (f RemoteSecretFilterFunc) Matches(ctx context.Context, matchable Matchable, remoteSecret *v1beta1.RemoteSecret) bool {
+	return f(ctx, matchable, remoteSecret)
+}
+
+var _ RemoteSecretFilter = (RemoteSecretFilterFunc)(nil)
+
+// DefaultRemoteSecretFilterFunc filters RemoteSecret based on three conditions: the SecretData must be obtained,
+// type of the secret must SecretTypeBasicAuth, and the RemoteSecret must have target in the same (local) namespace as matchable.
+var DefaultRemoteSecretFilterFunc = RemoteSecretFilterFunc(func(ctx context.Context, matchable Matchable, remoteSecret *v1beta1.RemoteSecret) bool {
```
Comment (metlos) [inline pkg/serviceprovider/rsfilter.go:42]: Do we anticipate any other way of matching the remote secret? Or rather will we ever have to choose from multiple ways of searching for remote secrets at runtime or is this only a way of simplifying the tests?

Comment (xbaran4) [reply pkg/serviceprovider/rsfilter.go:42]: I think the way we search for remotesecrets will change in the future but that probably won't mean there will be multiple filter strategies. So mostly just simplifying tests.

File: pkg/serviceprovider/clientbuilder.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (xbaran4) [reply pkg/serviceprovider/clientbuilder.go:26]: There is no final decision yet, but as I said, I would prefer to remove the `CredentialsReference` (that means leaving the CRD unchanged), and only after we know what and why is needed we might add it to the status.

**Review by mshaposhnik — APPROVED** [review]

File: pkg/serviceprovider/clientbuilder.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (metlos) [reply pkg/serviceprovider/clientbuilder.go:26]: 👍🏼 for the removal

**Comment by xbaran4** [comment]
Script test passed:
```shell
Checking access to public repositories without any tokens...
Checking repository: https://github.com/xbaran4/remote-secret
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-public-test" deleted
Checking repository: https://gitlab.com/xbaran4/public-test
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-public-test" deleted
Checking repository: https://quay.io/pbaran/rs
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-public-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-public-test" deleted
Done.

Creating SPIAccessTokens for each provider to test private repo...
spiaccesstoken.appstudio.redhat.com/test-access-token-github.com created
spiaccesstoken.appstudio.redhat.com/test-access-token-gitlab.com created
spiaccesstoken.appstudio.redhat.com/test-access-token-quay.io created
Done.

Checking access to private repositories with SPIAccessToken...
Checking repository: https://github.com/xbaran4/testing
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-private-test" deleted
Checking repository: https://gitlab.com/xbaran4/test
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-private-test" deleted
Checking repository: https://quay.io/pbaran/private-test
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-private-test" deleted
Done.

Removing SPIAccessTokens...
spiaccesstoken.appstudio.redhat.com "test-access-token-github.com" deleted
spiaccesstoken.appstudio.redhat.com "test-access-token-gitlab.com" deleted
spiaccesstoken.appstudio.redhat.com "test-access-token-quay.io" deleted
Done.

Creating RemoteSecrets and UploadSecrets for each provider to test private repo...
remotesecret.appstudio.redhat.com/test-remote-secret-github.com created
secret/test-remote-secret-secret created
remotesecret.appstudio.redhat.com/test-remote-secret-gitlab.com created
secret/test-remote-secret-secret created
remotesecret.appstudio.redhat.com/test-remote-secret-quay.io created
secret/test-remote-secret-secret created
Done.

Checking access to private repositories with RemoteSecret...
Checking repository: https://github.com/xbaran4/testing
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-private-test" deleted
Checking repository: https://gitlab.com/xbaran4/test
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-private-test" deleted
Checking repository: https://quay.io/pbaran/private-test
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test created
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com/spiaccesscheck-private-test condition met
spiaccesscheck.appstudio.redhat.com "spiaccesscheck-private-test" deleted
Done.

Removing RemoteSecrets...
remotesecret.appstudio.redhat.com "test-remote-secret-github.com" deleted
remotesecret.appstudio.redhat.com "test-remote-secret-gitlab.com" deleted
remotesecret.appstudio.redhat.com "test-remote-secret-quay.io" deleted
Done.

TEST: PASSED

```

**Pull request merged**